api/middleware/userContext.go

```diff
@@ -24,48 +31,91 @@ func NewUserContext() *userContext {
 // MiddlewareHandlerFunc returns the handler func used by the gin server when processing requests
 func (middleware *userContext) MiddlewareHandlerFunc() gin.HandlerFunc {
 	return func(c *gin.Context) {
-		xForwardedFor := c.Request.Header.Get("x-forwarded-for")
-		xRealIp := c.Request.Header.Get("x-real-ip")
-		remoteAddr := c.Request.RemoteAddr
-
-		log.Debug("user context",
-			"x-forwarded-for", xForwardedFor,
-			"x-real-ip", xRealIp,
-			"remote-addr", remoteAddr,
-		)
-
 		userAgent := c.Request.Header.Get("user-agent")
 
-		clientIp := parseIPHeader(xForwardedFor)
-		if clientIp == "" {
-			clientIp = parseIPHeader(xRealIp)
-		}
-		if clientIp == "" {
-			clientIp = parseIPHeader(remoteAddr)
-		}
+		// Be careful when using http headers for getting the real ip since it might
+		// depend on multiple factors: intermediate proxies (use information from proxies
+		// only if they are trustworthy), custom headers (like for cloudfare, nginx).
+		clientIP := getClientIP(c.Request)
 
 		c.Set(UserAgentKey, userAgent)
-		c.Set(UserIpKey, clientIp)
+		c.Set(UserIpKey, clientIP)
 		c.Next()
 	}
 }
 
-func parseIPHeader(header string) string {
+// if the server is exposed directly, RemoteAddr should contain the client ip
+// if the server is behind trusted proxies, the additional headers can be used (carefully)
+func getClientIP(r *http.Request) string {
+	xForwardedFor := r.Header.Get(xForwardedForHeader)
+	xRealIP := r.Header.Get(xRealIPHeader)
+	cfConnectingIP := r.Header.Get(cfConnectingIPHeader)
+	remoteAddr := r.RemoteAddr
+
+	// TODO: change log level to trace
+	log.Debug("user context",
+		"x-forwarded-for", xForwardedFor,
+		"x-real-ip", xRealIP,
+		"remote-addr", remoteAddr,
+	)
+
+	ip := parseHeader(cfConnectingIP)
```

Comment:
can we do the parsing of the header fields dependent on the config?
e.g if you set in config that you are using cloudflare, then check the cfConnectingIP?
same for nginx

Reply:
added config for custom headers